Refactors the emote code into emote datums.

lzimann opened the pull request:
I was bored and decided to touch the amazing emote code. 

- [x] Finishes brain emotes

- [x] Figure out a nice way to do the help emote.
- [x] Check if I didn't forget any of the emotes

lzimann committed:
Refactors the emote code into datums.

JJRcop commented:
sorry

lzimann committed:
Forgot some files.

lzimann committed:
I am stupid.

ExcessiveUseOfCobblestone commented:
how will [this](https://github.com/tgstation/tgstation/blob/f6db08af2e67c0a8933557f9f69bb9a5eb4fa6f0/code/game/objects/items/weapons/implants/implant_explosive.dm#L23) work

lzimann commented:
@ExcessiveUseOfCobblestone https://github.com/tgstation/tgstation/pull/22405/files#diff-166cb828d9485072490186d5e424c559R43
I'll have to change it to get all /mob/living after Antur's prgets merged but it should work(still haven't tested this part).

ChangelingRain commented on code/modules/mob/living/carbon/alien/emote.dm:
```diff
@@ -0,0 +1,31 @@
+	key = "roar"
+	key_third_person = "roars"
+	message_alien = "roars"
+	message_larva = "rofly roars"
```
softly

ExcessiveUseOfCobblestone replied:
scooby doo reference 🐶

ChangelingRain commented on code/modules/mob/living/carbon/alien/humanoid/emote.dm:
```diff
@@ -1,88 +0,0 @@
-		if (act == "hiss")
-			playsound(src.loc, "hiss", 40, 1, 1)
-
-		if (act == "deathgasp")
```
you need to preserve alien screeching somehow

lzimann replied:
I completly missed the sound actually, I'll figure out a way to do it.

MrPerson commented on code/datums/emotes.dm:
```diff
@@ -0,0 +1,90 @@
+	var/restraint_check = FALSE //Checks if the mob is restrained before performing the emote
+	var/muzzle_ignore = FALSE //Will only work if the emote is EMOTE_AUDIBLE
+	var/list/mob_type_allowed_typecache = list() //Types that are allowed to use that emote
+	var/list/mob_type_blacklist_typecache = list() //Types that are NOT allowed to use that emote
```
This might sound super ultra pandemic, but could you initialized the ones that are supposed to be strings as ""? The idea is to init them as what they're supposed to be so a bool would be TRUE/FALSE, a number might be 0, strings are "", etc

Cyberboss replied:
s/pandemic/pendantic

MrPerson commented on code/datums/emotes.dm:
```diff
@@ -0,0 +1,90 @@
+	if(is_type_in_typecache(user, mob_type_blacklist_typecache))
+		world << "blacklist"
+		return FALSE
+	return TRUE
```
Little reminder for you to remove these in case you forget

ExcessiveUseOfCobblestone commented:
On mobile so not entirely sure, but you'll probably want to do a pass by of #22370 if it gets merged as well.

lzimann committed:
Finishes brain emotes
Fixes simple animal attacks(when no damage)
Re-adds sounds to some alien emotes.

ChangelingRain commented on code/modules/mob/living/simple_animal/slime/emote.dm:
```diff
@@ -1,89 +1,60 @@
+
+/datum/emote/slime/mood/run_emote(mob/user, params)
+	. = ..()
+	var/mob/living/simple_animal/slime/S = user
```
if(.) do this stuff

lzimann replied:
doing if(.) will fail if the emote has no message, which is the case of all mood emotes.

ChangelingRain replied:
eh.

lzimann committed:
Adds the help emote and removes debug messages.

lzimann commented:
Well this is ready(to review at least).

optimumtact commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,424 @@
+	else if(!params)
+		var/custom_emote = copytext(sanitize(input("Choose an emote to display.") as text|null), 1, MAX_MESSAGE_LEN)
+		if(custom_emote && !check_invalid(custom_emote))
+			var/type = input("Is this a visible or hearable emote?") as null|anything in list("Visible", "Hearable")
```
this is a sleep so you need to check (!qdeleted) after

optimumtact replied:
actually scratch that, these will never be deleted will they

lzimann replied:
no, if they are getting deleted something is really, really wrong

optimumtact commented on code/modules/mob/living/carbon/emote.dm:
```diff
@@ -1,171 +1,64 @@
+	muzzle_ignore = TRUE
+	emote_type = EMOTE_AUDIBLE
+
+/datum/emote/living/carbon/gnarl
```
is this a real emote or was it mean to be snarl?

lzimann replied:
[has one for aliens aswell](https://github.com/tgstation/tgstation/blob/eace4eec148e8a366ee2fe6673f852b59a2e77b1/code/modules/mob/living/carbon/monkey/emote.dm#L18)

optimumtact reviewed (approved):
This is excellent, apart from minor nitpicks it LGTM

PKPenguin321 commented on code/datums/emotes.dm:
```diff
@@ -0,0 +1,98 @@
+/datum/emote/proc/select_message_type(mob/user)
+	. = message
+	if(!muzzle_ignore && user.is_muzzled() && emote_type == EMOTE_AUDIBLE)
+		return "makes a [pick("strong ", "weak ", "")]noise"
```
.

PKPenguin321 reviewed (commented):
RIP *aflap

lzimann commented:
aflap is fine, i would never remove such an amazing emote

lzimann committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into datumemotes

lzimann committed:
Adds a way to override emote_type(mainly for custom emote). Also moves the implant trigger to all livings instead of carbon only.

ChangelingRain commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -1,380 +1,130 @@
+	. = ..()
+	var/mob/living/carbon/human/H = user
+	if(.)
+		world << ". if"
```
debug!

ChangelingRain commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -1,380 +1,130 @@
+		world << ". if"
+		H.startTailWag()
+	else
+		world << "end"
```
more debug!

ChangelingRain commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -405,6 +155,7 @@
 	update_body()
 
 /mob/living/carbon/human/proc/OpenWings()
+	world << "open wings being called"
```
debug!

ChangelingRain commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -413,9 +164,11 @@
 	update_body()
 
 /mob/living/carbon/human/proc/CloseWings()
+	world << "close wings being called"
```
wow!

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,429 @@
+	. = ..()
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if((H.dna.features["wings"] != "None") && !("wings" in H.dna.species.mutant_bodyparts))
```
Should that second thing be negated?

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,429 @@
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if((H.dna.features["wings"] != "None") && !("wings" in H.dna.species.mutant_bodyparts))
+			H.OpenWings()
```
This should vary based on if the wings are open or not, so open wings flap closed then open again

lzimann replied:
wings being open or not depends on the mutant_bodyparts("wings" or "wingsopen")

ChangelingRain commented on code/datums/emotes.dm:
```diff
@@ -0,0 +1,99 @@
+
+/datum/emote/proc/can_run_emote(mob/user)
+	. = TRUE
+	if((user.stat && key != "deathgasp") || (user.status_flags & FAKEDEATH))
```
this should probably be a generic stat check var, like 

    var/stat_allowed = CONSCIOUS
    if(stat > stat_allowed)
    	return FALSE

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,429 @@
+
+/datum/emote/living/surrender/run_emote(mob/user, params)
+	. = ..()
+	if(!user.sleeping)
```
use the above idea for a stat check here

ChangelingRain replied:
oh yeah this should use if(.)

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,429 @@
+
+/datum/emote/living/flip/run_emote(mob/user, params)
+	. = ..()
+	if(!user.restrained() || !user.sleeping)
```
stat check plus use `restraint_check = TRUE`

lzimann replied:
Honestly I'm just killing the restraint check for *flip. Not really an issue if a mob is flipping while restrained.

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,429 @@
+
+/datum/emote/living/collapse/run_emote(mob/user, params)
+	. = ..()
+	user.Paralyse(2)
```
if(.) do this

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,429 @@
+	mob_type_allowed_typecache = list(/mob/living)
+	mob_type_blacklist_typecache = list(/mob/living/simple_animal/slime, /mob/living/brain)
+
+/datum/emote/living/aflap
```
subtype of flap, so it'll also flap wings
also maybe make the flap speed a var, so this is faster?

ChangelingRain commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -1,380 +1,130 @@
+/datum/emote/living/carbon/human/wing/run_emote(mob/user, params)
+	. = ..()
+	var/mob/living/carbon/human/H = user
+	if(findtext(select_message_type(user), "open"))
```
this should only run `if(.)`

lzimann committed:
Removes some leftover debug messages.
Aflap is now a child of flap, each having a wing opening/closing timer, also improved the code there a bit.
Adds a stat_allowed var(all emotes are default to CONSCIOUS).

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -113,22 +109,34 @@
 /datum/emote/living/flip/run_emote(mob/user, params)
 	. = ..()
-	if(!user.restrained() || !user.sleeping)
+	if(!user.sleeping)
```
with the conscious check ou don't need this; it should just be if(.)

lzimann replied:
it won't work because how I made to return if there's no message(probably my own fault but I don't really know what to do there instead)

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -113,22 +109,34 @@
 
 /datum/emote/living/flap/run_emote(mob/user, params)
 	. = ..()
 	if(ishuman(user))
```
oh yeah this should be if(. && ishuman(user))

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,437 @@
+	message_robot = "memes up and die."
+	message_alien = "lets out a waning guttural screech, green blood bubbling from its maw..."
+	message_larva = "lets out a sickly hiss of air and falls limply to the floor..."
+	message_monkey = "lets out a faint chimper as it collapses and stops moving..."
```
stat_allowed = UNCONSCIOUS

ChangelingRain commented on code/modules/mob/living/emote.dm:
```diff
@@ -1,344 +1,437 @@
+	key = "gasp"
+	key_third_person = "gasps"
+	message = "gasps!"
+	emote_type = EMOTE_AUDIBLE
```
stat_allowed = UNCONSCIOUS

ChangelingRain commented on code/modules/mob/living/carbon/alien/emote.dm:
```diff
@@ -0,0 +1,36 @@
+	key = "sign"
+	key_third_person = "signs"
+	message = "signs."
+	message_param = "signs the number %t."
```
These do limit parems to numbers, right?

lzimann replied:
uhh i might have forgotten to check for that

lzimann committed:
Adds a new proc, select_param, you can override it if you need specific checks in the param selection.
